### PR TITLE
Add Istio user agent to kube-client used in istioctl and other binaries

### LINKED ISSUE #28231: Istio binaries should set kubeclient UserAgent with version
At present istioctl and other binaries do not explicitly set kubeclient UserAgent.
Kubeclient-go guesses using binary name and puts 0.0.0 for version.

```
requestMetadata: {
callerIp: "35.192.190.99"
callerSuppliedUserAgent: "pilot-discovery/v0.0.0 (linux/amd64) kubernetes/$Format"
}
```

This information is useful during debugging.
`pkg.kube.util.go#SetRestDefaults()` is where it calls `rest.DefaultKubernetesUserAgent()` to set this property.

### PR DESCRIPTION
**Pull request by mandarjog**
Add istio-user-agent string to api server calls from istio for auditability.

resolves https://github.com/istio/istio/issues/28231

1.  Use `pkg/version` to generation version string for user-agent.
2. Consolidate multiple paths to build a kubeclient. 
3. Add klog support for client-go. The standard `-v=x` option cannot be used because we use it elsewhere. istioctl renames the option, and makes it a long form option with `--vklog`.

-- 

Following is output with `--vkog=9`. This flag was also added as part of this PR. shows `-H "User-Agent: istioctl/1.11-dev"` 

```console
jog@dev3:~/go/src/istio.io/istio$ ./out/linux_amd64/istioctl version -vklog=9
I0608 07:11:00.050509  250705 loader.go:372] Config loaded from file:  /home/mjog/.kube/config
I0608 07:11:00.051002  250705 loader.go:372] Config loaded from file:  /home/mjog/.kube/config
I0608 07:11:00.052499  250705 round_trippers.go:435] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: istioctl/1.11-dev" 'https://34.67.144.238/api/v1/namespaces/istio-system/pods?fieldSelector=status.phase%3DRunning&labelSelector=app%3Distiod&timeout=10m0s'
I0608 07:11:00.180675  250705 round_trippers.go:454] GET https://34.67.144.238/api/v1/namespaces/istio-system/pods?fieldSelector=status.phase%3DRunning&labelSelector=app%3Distiod&timeout=10m0s 200 OK in 128 milliseconds
I0608 07:11:00.180739  250705 round_trippers.go:460] Response Headers:
I0608 07:11:00.180774  250705 round_trippers.go:463]     Cache-Control: no-cache, private
I0608 07:11:00.180805  250705 round_trippers.go:463]     Content-Type: application/json
I0608 07:11:00.180834  250705 round_trippers.go:463]     Date: Tue, 08 Jun 2021 07:11:00 GMT
I0608 07:11:00.180838  250705 round_trippers.go:463]     Audit-Id: 1c8eebbc-0860-4ac0-8f9e-278b54c8bd8c
I0608 07:11:00.213594  250705 request.go:1123] Response Body {...}
```

### CODE REVIEW

File: operator/cmd/mesh/upgrade.go [hunk 1]
```diff
@@ -443,7 +444,7 @@ type ExecClient interface {
 
 // NewClient is the constructor for the client wrapper
 func NewClient(kubeconfig, configContext string) (*Client, error) {
-	config, err := defaultRestConfig(kubeconfig, configContext)
+	config, err := kube.DefaultRestConfig(kubeconfig, configContext)
```
Comment (howardjohn) [inline operator/cmd/mesh/upgrade.go:447]: seems simpler change but should it be in another PR?

Comment (howardjohn) [reply operator/cmd/mesh/upgrade.go:447]: Oh nevermind, i see we are setting user agent at a very low level

**Review by howardjohn — COMMENTED** [review]
looks good

File: istioctl/cmd/istioctl/main.go [hunk 2]
```diff
@@ -33,8 +37,28 @@ func main() {
...
 	if err := rootCmd.Execute(); err != nil {
 		exitCode := cmd.GetExitCode(err)
 		os.Exit(exitCode)
 	}
 }
+
+// EnableKlogWithCobra enables klog to work with cobra / pflags.
+// k8s libraries like client-go use klog.
+func EnableKlogWithCobra() {
+	klog.InitFlags(nil)
```
Comment (howardjohn) [inline istioctl/cmd/istioctl/main.go:51]: does passing nil (which then uses goflag.CommandLine) cause any issues here in help messages, etc? Should we create a dedicated FlagSet for this so it doesn't unexpectedly leak out?

File: istioctl/cmd/istioctl/main.go [hunk 3]
```diff
@@ -33,8 +37,28 @@ func main() {
...
+	gf := goflag.CommandLine.Lookup("v")
+	pflag.CommandLine.AddFlag(pflag.PFlagFromGoFlag(
+		&goflag.Flag{
+			Name:     "vklog",
+			Value:    gf.Value,
+			DefValue: gf.DefValue,
+			Usage:    gf.Usage + ". Like -v flag. ex: --vklog=9",
+		}))
+
+	klog.SetLogger(nil)
```
Comment (howardjohn) [inline istioctl/cmd/istioctl/main.go:63]: If I understand the situation correctly:

Today we can enable "debug" logs from klog with  `--log_output_level=klog:debug`. This uses our logging system (json, stackdriver, etc). But it uses the default klog verbosity level which is not 9.

This pr allows setting the klog verbosity, but it doesn't use our logging system. I think if you remove this line it will allow --vklog to work, but it also requires --log_output_level=klog:debug. It seems preferable to use our logging system, but maybe we can improve the UX

I feel like this is something that is generally useful outside of istioctl - I have often wanted this in istiod. Maybe we should make it so it can be configured as part of istio/pkg? Not sure the UX we want here though (new flag, --log_output_level=klog:debug for most cases and env var to tune -v level, etc).

**Review by howardjohn — COMMENTED** [review]
LGTM except comments on the vklog stuff. I think its awesome but may need some tweaks

File: istioctl/cmd/istioctl/main.go (hunk 3 above)
Comment (mandarjog) [reply istioctl/cmd/istioctl/main.go:63]: yeah, once we have it in istioctl, we will add it to add binaries. 
In a previous version of the PR I had enabled it by mistake on istiod and is spews logs, but it is very useful.

Comment (howardjohn) [reply istioctl/cmd/istioctl/main.go:63]: Can we at least remove `	klog.SetLogger(nil)` so it uses our logging system?

**Review by howardjohn — APPROVED** [review]

**Comment by mandarjog** [comment]
@howardjohn I removed a bunch of code and now setting `--vklog` will correctly adjust the loglevel for klog so the user do not need to provide additional `--log_output_level=klog:debug`.
I will move all this into `istio.io/pkg` soon.

**Comment by mandarjog** [comment]
@esnible @therealmitchconnors for UX.

**Review by howardjohn — APPROVED** [review]

**Review by esnible — APPROVED** [review]
